com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationArchiveExtractor.java

```diff
@@ -0,0 +1,200 @@
+package com.sap.cloud.lm.sl.cf.process.util;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.text.MessageFormat;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+import org.apache.commons.io.FilenameUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.sap.cloud.lm.sl.common.ContentException;
+import com.sap.cloud.lm.sl.common.SLException;
+
+public class ApplicationArchiveExtractor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ApplicationArchiveExtractor.class);
+
+    static final String PATH_SHOULD_NOT_BE_ABSOLUTE = "Archive entry name \"{0}\" should not be absolute";
+    static final String PATH_SHOULD_BE_NORMALIZED = "Archive entry name \"{0}\" should be normalized";
+    private static final int BUFFER_SIZE = 4 * 1024; // 4KB
+    private static final String ARCHIVE_FILE_SEPARATOR = "/";
+    private InputStream appArchiveStream;
+    private String fileName;
+
+    public ApplicationArchiveExtractor(InputStream appArchiveStream, String fileName) {
+        this.appArchiveStream = appArchiveStream;
+        this.fileName = fileName;
+    }
+
+    public Path extract(ExtractStatusCallback callback) throws IOException, SLException {
+        if (callback == null) {
+            callback = ExtractStatusCallback.NONE;
+        }
+
+        try (InputStreamProducer streamProducer = new InputStreamProducer(appArchiveStream, fileName)) {
+            return extract(streamProducer, callback);
+        } catch (IOException | SLException e) {
+            callback.onError(e);
+            throw e;
+        } catch (Exception e) {
+            callback.onError(e);
+            return null;
+        }
+    }
+
+    private Path extract(InputStreamProducer streamProducer, ExtractStatusCallback callback) throws IOException, SLException {
```

Comment:
This method is not symmetric to the one above it. The first uses a field (appArchiveStream) while this one uses a method argument. Either use a method argument in both methods, or a field.